### ISSUE TITLE
Request: Equivalent of theano.clone() to dynamically replace subgraph

### ISSUE THREAD

**Issue by sotelo** [tensorflow/tensorflow#1070]
I would like to know if this feature exist in the library or if there's any plan to have it in the future.

Thanks!

**Comment by vrv** [comment]
Hopefully someone else knows more about the literature, but is "weight noise" a generalized stochastic operation applied specifically to weights?  For example, is dropout a form of "weight noise" if applied to weights?

You can implement dropout as a composition of existing ops here: https://github.com/tensorflow/tensorflow/blob/8048088bfb82846078f8023bc6199e6424926624/tensorflow/python/ops/nn_ops.py#L581, so if you can express the specific noise algorithm you want as a composition of existing ops, then you can do that today.  If there's something more specific you are asking for, let us know, and feel free to contribute it to the library.

**Comment by sotelo** [comment]
Thanks for answering :)

Dropout it's very similar to what I would like to do. The only problem is that I would like to apply it AFTER the rest of the computation graph is defined. As far as I understand, the current dropout function has to be applied before. This is what I am referring to when I say graph replacement. In theano, the mechanism to do this is use the theano.clone function with a replace dictionary as argument. The reason that I would like to do it like that, is that I want to use this kind of function in all the weights. So, for example, the weights that correspond to a LSTM that I don't have access to. Ideally, I would like to do something like this:

vars = tf.variables.all_variables()
new_vars = [tf.nn.dropout(var) for var in vars ]
tf.replace(vars, new_vars)

For a less general example: How can I apply dropout to the weights of a lstm?

I saw this discussion that is related, but I would like to know if there is a simpler way of doing things:
http://stackoverflow.com/questions/33748552/tensorflow-how-to-replace-a-node-in-a-calculation-graph

Thanks!

**Comment by vrv** [comment]
OK great, I've renamed the title to reflect the more general feature request.

I can think of various ways to do this, but none of them are the same as theano.clone. They involve duplicating the entire graph as mentioned in the SO post, or decomposing your graph into "functions" (a currently experimental and evolving feature) and using tf.cond to select which branch to run at runtime, etc, or using sess.partial_run or tf.py_func (also both evolving and experimental) to embed the logic in the user program rather than in the graph.

Thanks for the feature request!

**Comment by rdipietro** [comment]
+1 :)

**Comment by girving** [comment]
Closing since the merged PR #557 seems related.
